pkg/endpoint/bpf.go

```diff
@@ -526,234 +526,238 @@ func (e *Endpoint) regenerateBPF(owner Owner, regenContext *regenerationContext)
 	stats := &regenContext.Stats
 	stats.waitingForLock.Start()
 
+	datapathRegenCtxt := regenContext.datapathRegenerationContext
+
 	// Make sure that owner is not compiling base programs while we are
 	// regenerating an endpoint.
 	owner.GetCompilationLock().RLock()
+	stats.waitingForLock.End(true)
 	defer owner.GetCompilationLock().RUnlock()
 
-	err = e.LockAlive()
-	stats.waitingForLock.End(err == nil)
-	if err != nil {
-		return 0, compilationExecuted, err
-	}
+	datapathRegenCtxt.prepareForProxyUpdates()
+	defer datapathRegenCtxt.completionCancel()
 
-	datapathRegenCtxt := regenContext.datapathRegenerationContext
-	datapathRegenCtxt.prepareForDatapathRegeneration()
-
-	currentDir := datapathRegenCtxt.currentDir
-	nextDir := datapathRegenCtxt.nextDir
-
-	// In the first ever regeneration of the endpoint, the conntrack table
-	// is cleaned from the new endpoint IPs as it is guaranteed that any
-	// pre-existing connections using that IP are now invalid.
-	if !e.ctCleaned {
-		go func() {
-			ipv4 := !option.Config.IPv4Disabled
-			created := ctmap.Exists(nil, ipv4, true)
-			if e.ConntrackLocal() {
-				created = ctmap.Exists(e, ipv4, true)
-			}
-			if created {
-				e.scrubIPsInConntrackTable()
-			}
-			close(datapathRegenCtxt.ctCleaned)
-		}()
-	} else {
-		close(datapathRegenCtxt.ctCleaned)
-	}
+	err = func() error {
```

Comment:
Since this function is so long, a comment explaining what it does and why it's a function-in-a-function would be good (e.g. that it depends on drymode). I'm guessing this is a step towards pulling it all out of here, but if it was all just to ensure the .Unlock is deferred I'm on the fence about it. It's definitely better, but is it too weird a thing to see in the code, especially because it's so long.

Reply:
Ah, you do factor it out!